sky/utils/controller_utils.py

```diff
@@ -679,26 +696,56 @@ def maybe_translate_local_file_mounts_and_sync_up(task: 'task_lib.Task',
             ux_utils.spinner_message(
                 f'Translating {msg} to SkyPilot Storage...'))
 
+    # Get the bucket name for the workdir and file mounts,
+    # we store all these files in same bucket from config.
+    bucket_wth_prefix = skypilot_config.get_nested(('jobs', 'bucket'), None)
+    store_kwargs: Dict[str, Any] = {}
+    if bucket_wth_prefix is None:
+        store_type = store_cls = sub_path = None
+        storage_account_name = region = None
+        bucket_name = constants.FILE_MOUNTS_BUCKET_NAME.format(
+            username=common_utils.get_cleaned_username(), id=run_id)
+    else:
+        store_type, store_cls, bucket_name, sub_path, storage_account_name, \
+            region = storage_lib.StoreType.get_fields_from_store_url(
+                bucket_wth_prefix)
+        if storage_account_name is not None:
+            store_kwargs['storage_account_name'] = storage_account_name
+        if region is not None:
+            store_kwargs['region'] = region
+
     # Step 1: Translate the workdir to SkyPilot storage.
     new_storage_mounts = {}
     if task.workdir is not None:
-        bucket_name = constants.WORKDIR_BUCKET_NAME.format(
-            username=common_utils.get_cleaned_username(), id=run_id)
         workdir = task.workdir
         task.workdir = None
         if (constants.SKY_REMOTE_WORKDIR in original_file_mounts or
                 constants.SKY_REMOTE_WORKDIR in original_storage_mounts):
             raise ValueError(
                 f'Cannot mount {constants.SKY_REMOTE_WORKDIR} as both the '
                 'workdir and file_mounts contains it as the target.')
-        new_storage_mounts[
-            constants.
-            SKY_REMOTE_WORKDIR] = storage_lib.Storage.from_yaml_config({
-                'name': bucket_name,
-                'source': workdir,
-                'persistent': False,
-                'mode': 'COPY',
-            })
+        bucket_sub_path = _sub_path_join(
+            sub_path,
+            constants.FILE_MOUNTS_WORKDIR_SUBPATH.format(run_id=run_id))
+        stores = None
+        if store_type is not None:
+            assert store_cls is not None
+            with sky_logging.silent():
+                stores = {
+                    store_type: store_cls(name=bucket_name,
+                                          source=workdir,
+                                          _bucket_sub_path=bucket_sub_path,
+                                          **store_kwargs)
+                }
+                assert_no_bucket_creation(stores[store_type])
```

Comment:
Do we need to wrap this in a try catch block and raise a cleaner error? E.g., consider the case when an admin created a bucket, but did not grant get/list permissions on the bucket to the user running managed jobs. In that case, the store initialization will fail with a permission denied error - we may want to raise that clearly here.

Reply:
On a related note, @zpoint can you help manually test the scenario where an admin grants limited permissions to a user to access a bucket and make sure this PR works for that case?

Here's a suggestion how to do it:

1. Create a new AWS IAM user (say `user1`) which does not have permissions to create new buckets.
2. With an admin AWS account, create a S3 bucket `bucket1` and grant access to `user1`
3. With `user1` credentials configured in your local SkyPilot installation, try using this managed jobs PR with `bucket1` as the intermediate bucket

We should verify both:
1. Sanity check: `user1` is not able to use regular `sky jobs launch` with a `workdir` on master branch because they do not have permissions to create bucket.
2. With this branch and `bucket1` configured in `config.yaml`, `user1` is able to use `sky jobs launch`

Reply:
1. No bucket configured, no permission
```
(sky) ➜  skypilot git:(dev/zeping/specific_bucket_for_sky_jobs) sky jobs launch ~/Desktop/hello-sky/work_dir_1.yaml -y
Task from YAML spec: /Users/zepingguo/Desktop/hello-sky/work_dir_1.yaml
  Created AZ Container 'zpoint-filemounts-bucket' in 'eastus' under storage account 'sky635694a6141885a3'.
  Excluded files to sync to cluster based on .gitignore.
✓ Storage synced: ~/Desktop/dir1 -> https://sky635694a6141885a3.blob.core.windows.net/zpoint-filemounts-bucket/  View logs at: ~/sky_logs/sky-2024-12-26-20-45-00-084283/storage_sync.log
Resources for managed job 'test_workdir_bucket_name_1' will be computed on the managed jobs controller, since --yes is set.
⚙︎ Translating workdir and file_mounts with local source paths to SkyPilot Storage...
  Workdir: '.' -> storage: 'skypilot-filemounts-zepingguo-8e20d477'.
  Folder : '~/Desktop/dir1' -> storage: 'skypilot-filemounts-zepingguo-8e20d477'.
  Folder : '~/Desktop/dir2' -> storage: 'skypilot-filemounts-zepingguo-8e20d477'.
  Files (listed below)  -> storage: skypilot-filemounts-zepingguo-8e20d477:
    ~/Desktop/dir1/dir1.py
Could not create StoreType.S3 store with name skypilot-filemounts-zepingguo-8e20d477.
IndexError: list index out of range
```

2. Configured no premission bucket

```
(sky) ➜  skypilot git:(dev/zeping/specific_bucket_for_sky_jobs) sky jobs launch ~/Desktop/hello-sky/work_dir_1.yaml -y
Task from YAML spec: /Users/zepingguo/Desktop/hello-sky/work_dir_1.yaml
Verifying bucket and uploading from source for storage zpoint-filemounts-bucket
  Excluded files to sync to cluster based on .gitignore.
✓ Storage synced: ~/Desktop/dir1 -> https://sky635694a6141885a3.blob.core.windows.net/zpoint-filemounts-bucket/  View logs at: ~/sky_logs/sky-2024-12-26-20-47-59-674784/storage_sync.log
Storage type StoreType.AZURE already exists under storage account 'sky635694a6141885a3'.
Resources for managed job 'test_workdir_bucket_name_1' will be computed on the managed jobs controller, since --yes is set.
⚙︎ Translating workdir and file_mounts with local source paths to SkyPilot Storage...
botocore.exceptions.ClientError: An error occurred (403) when calling the HeadBucket operation: Forbidden

The above exception was the direct cause of the following exception:

sky.exceptions.StorageBucketGetError: Failed to access existing bucket 'zpoint-bucket-s3-intermediate'. This is likely because it is a private bucket you do not have access to.
To fix: 
  1. If you are trying to create a new bucket: use a different name.
  2. If you are trying to connect to an existing bucket: make sure your cloud credentials have access to it. To debug, consider running `aws s3 ls zpoint-bucket-s3-intermediate`.
 ```

3. Configured bucket with permission

```
(sky) ➜  skypilot git:(dev/zeping/specific_bucket_for_sky_jobs) sky jobs launch ~/Desktop/hello-sky/work_dir_1.yaml -y
Task from YAML spec: /Users/zepingguo/Desktop/hello-sky/work_dir_1.yaml
Verifying bucket and uploading from source for storage zpoint-filemounts-bucket
  Excluded files to sync to cluster based on .gitignore.
✓ Storage synced: ~/Desktop/dir1 -> https://sky635694a6141885a3.blob.core.windows.net/zpoint-filemounts-bucket/  View logs at: ~/sky_logs/sky-2024-12-26-21-00-02-263963/storage_sync.log
Storage type StoreType.AZURE already exists under storage account 'sky635694a6141885a3'.
Resources for managed job 'test_workdir_bucket_name_1' will be computed on the managed jobs controller, since --yes is set.
⚙︎ Translating workdir and file_mounts with local source paths to SkyPilot Storage...
  Workdir: '.' -> storage: 'bucket-admin-test'.
  Folder : '~/Desktop/dir1' -> storage: 'bucket-admin-test'.
  Folder : '~/Desktop/dir2' -> storage: 'bucket-admin-test'.
  Files (listed below)  -> storage: bucket-admin-test:
    ~/Desktop/dir1/dir1.py
  Excluded files to sync to cluster based on .gitignore.
✓ Storage synced: ~/Desktop/dir1 -> https://sky635694a6141885a3.blob.core.windows.net/zpoint-filemounts-bucket/  View logs at: ~/sky_logs/sky-2024-12-26-21-00-11-041926/storage_sync.log
'.git' directory under '.' is excluded during sync.
  Excluded files to sync to cluster based on .gitignore.
✓ Storage synced: . -> s3://bucket-admin-test/  View logs at: ~/sky_logs/sky-2024-12-26-21-00-14-963562/storage_sync.log
  Excluded files to sync to cluster based on .gitignore.
✓ Storage synced: ~/Desktop/dir1 -> s3://bucket-admin-test/  View logs at: ~/sky_logs/sky-2024-12-26-21-01-17-444519/storage_sync.log
  Excluded files to sync to cluster based on .gitignore.
✓ Storage synced: ~/Desktop/dir2 -> s3://bucket-admin-test/  View logs at: ~/sky_logs/sky-2024-12-26-21-01-20-386086/storage_sync.log
  Excluded files to sync to cluster based on .gitignore.
✓ Storage synced: /var/folders/83/zxqx914s57x310rfnhq8kk9r0000gn/T/skypilot-filemounts-files-ee46694f -> s3://bucket-admin-test/  View logs at: ~/sky_logs/sky-2024-12-26-21-01-23-026236/storage_sync.log
✓ Uploaded local files/folders.
Launching managed job 'test_workdir_bucket_name_1' from jobs controller...
Choosing resources for managed jobs controller...
Considered resources (1 node):
------------------------------------------------------------------------------------------
 CLOUD   INSTANCE      vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
------------------------------------------------------------------------------------------
 AWS     m6i.2xlarge   8       32        -              us-east-1     0.38          ✔     
------------------------------------------------------------------------------------------
⚙︎ Launching managed jobs controller on AWS us-east-1 (us-east-1a,us-east-1b,us-east-1c,us-east-1d,us-east-1f).
└── Instance is up.
✓ Cluster launched: sky-jobs-controller-94a61418.  View logs at: ~/sky_logs/sky-2024-12-26-21-01-29-148802/provision.log
⚙︎ Mounting files.
  Syncing (to 1 node): /var/folders/83/zxqx914s57x310rfnhq8kk9r0000gn/T/managed-dag-test_workdir_bucket_name_1-yh2u7hqc -> ~/.sky/managed_jobs/test_workdir_bucket_name_1-2906.yaml
  Syncing (to 1 node): /var/folders/83/zxqx914s57x310rfnhq8kk9r0000gn/T/tmp3owt4r01 -> ~/.sky/managed_jobs/test_workdir_bucket_name_1-2906.config_yaml
✓ Files synced.  View logs at: ~/sky_logs/sky-2024-12-26-21-01-29-148802/file_mounts.log
⚙︎ Running setup on managed jobs controller.
  Check & install cloud dependencies on controller: done.                                        
✓ Setup completed.  View logs at: ~/sky_logs/sky-2024-12-26-21-01-29-148802/setup-*.log
⚙︎ Job submitted, ID: 1
⠹ Waiting for task to start (status: PENDING). It may take a few minutes.
  View controller logs: sky jobs logs --controller 1
```

@romilbhardwaj It's working, but I'm not sure if the exception hint is clear enough at the moment. @romilbhardwaj

Reply:
Awesome, thanks for testing @zpoint! Case 2 and 3 look good, but for case 1 can we have a better exception saying Permission Denied or something similar to case 2?

Reply:
What about now @romilbhardwaj 

case 1 No bucket configured, no permission:

```
(sky) ➜  skypilot git:(dev/zeping/specific_bucket_for_sky_jobs) ✗ sky jobs launch ~/Desktop/hello-sky/work_dir_1.yaml -y
Task from YAML spec: /Users/zepingguo/Desktop/hello-sky/work_dir_1.yaml
Resources for managed job 'test_workdir_bucket_name_1' will be computed on the managed jobs controller, since --yes is set.
⚙︎ Translating workdir and file_mounts with local source paths to SkyPilot Storage...
  Workdir: '.' -> storage: 'skypilot-filemounts-zepingguo-d4826954'.
  Folder : '~/Desktop/dir1' -> storage: 'skypilot-filemounts-zepingguo-d4826954'.
  Folder : '~/Desktop/dir2' -> storage: 'skypilot-filemounts-zepingguo-d4826954'.
  Files (listed below)  -> storage: skypilot-filemounts-zepingguo-d4826954:
    ~/Desktop/dir1/dir1.py
Could not create StoreType.S3 store with name skypilot-filemounts-zepingguo-d4826954.
Error setting storage status: Storage skypilot-filemounts-zepingguo-d4826954 not found.
botocore.exceptions.ClientError: An error occurred (AccessDenied) when calling the CreateBucket operation: User: arn:aws:iam::195275664570:user/user1 is not authorized to perform: s3:CreateBucket on resource: "arn:aws:s3:::skypilot-filemounts-zepingguo-d4826954" with an explicit deny in an identity-based policy

The above exception was the direct cause of the following exception:

sky.exceptions.StorageBucketCreateError: Attempted to create a bucket skypilot-filemounts-zepingguo-d4826954 but failed.
```

Reply:
This looks good, thanks!